File: AssetManager/Data/DataFunctions/ClientSideFunctions/AttribIndexFunctions.vb
Module AttribIndexFunctions

    Public Sub FillComboBox(IndexType() As AttributeDataStruct, ByRef cmb As ComboBox)
        cmb.Items.Clear()
        cmb.Text = ""
        For Each ComboItem As AttributeDataStruct In IndexType
            cmb.Items.Add(ComboItem.DisplayValue)
        Next
    End Sub

    Public Sub FillToolComboBox(IndexType() As AttributeDataStruct, ByRef cmb As ToolStripComboBox)
        cmb.Items.Clear()
        cmb.Text = ""
        Dim i As Integer = 0
        For Each ComboItem As AttributeDataStruct In IndexType
            cmb.Items.Insert(i, ComboItem.DisplayValue)
            i += 1
        Next
    End Sub

    Public Function GetDBValue(codeIndex() As AttributeDataStruct, index As Integer) As String
        Try
            If index > -1 Then
                Return codeIndex(index).Code
            End If
            Return String.Empty
        Catch
            Return String.Empty
        End Try
    End Function

    Public Function GetDisplayValueFromCode(codeIndex() As AttributeDataStruct, code As String) As String
        For Each item In codeIndex
            If item.Code = code Then Return item.DisplayValue
        Next
        Return String.Empty
    End Function

    Public Function GetDisplayValueFromIndex(codeIndex() As AttributeDataStruct, index As Integer) As String
        Return codeIndex(index).DisplayValue
    End Function

    Public Function GetComboIndexFromCode(codeIndex() As AttributeDataStruct, code As String) As Integer
        For i = 0 To codeIndex.Length - 1
            If codeIndex(i).Code = code Then Return i
        Next
        Return -1
    End Function

    Public Sub PopulateAttributeIndexes()
        Dim BuildIdxs = Task.Run(Sub()
                                     DeviceAttribute.Locations = BuildIndex(DevicesBaseCols.AttribTable, DeviceAttribType.Location)
                                     DeviceAttribute.ChangeType = BuildIndex(DevicesBaseCols.AttribTable, DeviceAttribType.ChangeType)
                                     DeviceAttribute.EquipType = BuildIndex(DevicesBaseCols.AttribTable, DeviceAttribType.EquipType)
                                     DeviceAttribute.OSType = BuildIndex(DevicesBaseCols.AttribTable, DeviceAttribType.OSType)
                                     DeviceAttribute.StatusType = BuildIndex(DevicesBaseCols.AttribTable, DeviceAttribType.StatusType)
                                     SibiAttribute.StatusType = BuildIndex(SibiRequestCols.AttribTable, SibiAttribType.SibiStatusType)
                                     SibiAttribute.ItemStatusType = BuildIndex(SibiRequestCols.AttribTable, SibiAttribType.SibiItemStatusType)
                                     SibiAttribute.RequestType = BuildIndex(SibiRequestCols.AttribTable, SibiAttribType.SibiRequestType)
                                     SibiAttribute.AttachFolder = BuildIndex(SibiRequestCols.AttribTable, SibiAttribType.SibiAttachFolder)
                                 End Sub)
        BuildIdxs.Wait()
    End Sub

    Public Function BuildIndex(codeType As String, typeName As String) As AttributeDataStruct()
        Try
            Dim strQRY = "SELECT * FROM " & codeType & " LEFT OUTER JOIN munis_codes on " & codeType & ".db_value = munis_codes.asset_man_code WHERE type_name ='" & typeName & "' ORDER BY " & ComboCodesBaseCols.DisplayValue & ""
            Using results As DataTable = DBFactory.GetDatabase.DataTableFromQueryString(strQRY)
                Dim tmpArray As New List(Of AttributeDataStruct)
                For Each r As DataRow In results.Rows
                    Dim DisplayValue As String = ""
                    If r.Table.Columns.Contains("munis_code") Then
                        If Not IsDBNull(r.Item("munis_code")) Then
                            DisplayValue = r.Item(ComboCodesBaseCols.DisplayValue).ToString & " - " & r.Item("munis_code").ToString
                        Else
                            DisplayValue = r.Item(ComboCodesBaseCols.DisplayValue).ToString
                        End If
                    Else
                        DisplayValue = r.Item(ComboCodesBaseCols.DisplayValue).ToString
                    End If
                    tmpArray.Add(New AttributeDataStruct(DisplayValue, r.Item(ComboCodesBaseCols.CodeValue).ToString, CInt(r.Item(ComboCodesBaseCols.ID))))
                Next
                Return tmpArray.ToArray
            End Using
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
            Return Nothing
        End Try
    End Function

End Module

File: AssetManager/Data/DataFunctions/ClientSideFunctions/AttribIndexFunctions.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.Linq;
using System.Threading.Tasks;
namespace AssetManager
{
    static class AttribIndexFunctions
    {

        public static void FillComboBox(AttributeDataStruct[] IndexType, ComboBox cmb)
        {
            cmb.Items.Clear();
            cmb.Text = "";
            foreach (AttributeDataStruct ComboItem in IndexType)
            {
                cmb.Items.Add(ComboItem.DisplayValue);
            }
        }

        public static void FillToolComboBox(AttributeDataStruct[] IndexType, ref ToolStripComboBox cmb)
        {
            cmb.Items.Clear();
            cmb.Text = "";
            int i = 0;
            foreach (AttributeDataStruct ComboItem in IndexType)
            {
                cmb.Items.Insert(i, ComboItem.DisplayValue);
                i += 1;
            }
        }

        public static string GetDBValue(AttributeDataStruct[] codeIndex, int index)
        {
            try
            {
                if (index > -1)
                {
                    return codeIndex[index].Code;
                }
                return string.Empty;
            }
            catch
            {
                return string.Empty;
            }
        }

        public static string GetDisplayValueFromCode(AttributeDataStruct[] codeIndex, string code)
        {
            foreach (AttributeDataStruct item in codeIndex)
            {
                if (item.Code == code)
                    return item.DisplayValue;
            }
            return string.Empty;
        }

        public static string GetDisplayValueFromIndex(AttributeDataStruct[] codeIndex, int index)
        {
            return codeIndex[index].DisplayValue;
        }

        public static int GetComboIndexFromCode(AttributeDataStruct[] codeIndex, string code)
        {
            for (int i = 0; i <= codeIndex.Length - 1; i++)
            {
                if (codeIndex[i].Code == code)
                    return i;
            }
            return -1;
        }

        public static void PopulateAttributeIndexes()
        {
            var BuildIdxs = Task.Run(() =>
            {
                GlobalInstances.DeviceAttribute.Locations = BuildIndex(DevicesBaseCols.AttribTable, DeviceAttribType.Location);
                GlobalInstances.DeviceAttribute.ChangeType = BuildIndex(DevicesBaseCols.AttribTable, DeviceAttribType.ChangeType);
                GlobalInstances.DeviceAttribute.EquipType = BuildIndex(DevicesBaseCols.AttribTable, DeviceAttribType.EquipType);
                GlobalInstances.DeviceAttribute.OSType = BuildIndex(DevicesBaseCols.AttribTable, DeviceAttribType.OSType);
                GlobalInstances.DeviceAttribute.StatusType = BuildIndex(DevicesBaseCols.AttribTable, DeviceAttribType.StatusType);
                GlobalInstances.SibiAttribute.StatusType = BuildIndex(SibiRequestCols.AttribTable, SibiAttribType.SibiStatusType);
                GlobalInstances.SibiAttribute.ItemStatusType = BuildIndex(SibiRequestCols.AttribTable, SibiAttribType.SibiItemStatusType);
                GlobalInstances.SibiAttribute.RequestType = BuildIndex(SibiRequestCols.AttribTable, SibiAttribType.SibiRequestType);
                GlobalInstances.SibiAttribute.AttachFolder = BuildIndex(SibiRequestCols.AttribTable, SibiAttribType.SibiAttachFolder);
            });
            BuildIdxs.Wait();
        }

        public static AttributeDataStruct[] BuildIndex(string codeType, string typeName)
        {
            try
            {
                var strQRY = "SELECT * FROM " + codeType + " LEFT OUTER JOIN munis_codes on " + codeType + ".db_value = munis_codes.asset_man_code WHERE type_name ='" + typeName + "' ORDER BY " + ComboCodesBaseCols.DisplayValue + "";
                using (DataTable results = AssetManager.DBFactory.GetDatabase().DataTableFromQueryString(strQRY))
                {
                    List<AttributeDataStruct> tmpArray = new List<AttributeDataStruct>();
                    foreach (DataRow r in results.Rows)
                    {
                        string DisplayValue = "";
                        if (r.Table.Columns.Contains("munis_code"))
                        {
                            if (!Information.IsDBNull(r["munis_code"]))
                            {
                                DisplayValue = r[ComboCodesBaseCols.DisplayValue].ToString() + " - " + r["munis_code"].ToString();
                            }
                            else
                            {
                                DisplayValue = r[ComboCodesBaseCols.DisplayValue].ToString();
                            }
                        }
                        else
                        {
                            DisplayValue = r[ComboCodesBaseCols.DisplayValue].ToString();
                        }
                        tmpArray.Add(new AttributeDataStruct(DisplayValue, r[ComboCodesBaseCols.CodeValue].ToString(), Convert.ToInt32(r[ComboCodesBaseCols.ID])));
                    }
                    return tmpArray.ToArray();
                }
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
                return null;
            }
        }

    }
}
